README
Module  : Sphinx Search
Version : 1.1.2
Author  : Thomas Seiffert, updated by Martijn van Maasakkers <martijn@vanmaasakkers.net>, Elan Ruusamäe <glen@delfi.ee>

Module for using sphinx as fulltext search-engine instead of phorum built-in search
sphinx can be downloaded and installed from http://sphinxsearch.com/

Installation instructions:
--------------------------

To get this to work you will need to install sphinx on your server.
This will not be possible in a shared hosting environment and no, there
is no workaround for that.

some caveats:
  - data is stored in both phorum_search-table (not needed here) and in the
    sphinx filesystem - this doubles the sizes needed for search even though
    the search-table is not needed. You can clear it out every now and then.
    you may disable filling phorum_search table by setting
    $PHORUM['DBCONFIG']['empty_search_table'] to true in db config.
    but then you need to rebuild search data if you want to go back to mysql
    fulltext search. you can rebuild search data using admin panel.
  - search-data is not updated in real-time - sphinx reindexes the data if the indexer
    is called for the index, which does a full reindexing of the sphinx index.
    There is a workaround available in sphinx http://sphinxsearch.com/doc.html#live-updates
    which works essentially with two indexes. A large one and smaller one with only the deltas
    to the full index and which can be reindexed therefore more often.
    Still, its not realtime.
    For me the full indexing of 350,000 messages took around 5 minutes.
    I'd run the full indexing around once a day with those times.
    See crontab file for sample crontab setup, install it to /etc/cron.d/phorum-sphinx-search

Before you can run a search through sphinx you will need to setup sphinx correctly and do at least one full
indexing of your phorum-data.
Try the searching on the command-line using search from the sphinx install first.
Also you will need to start searchd from sphinx so that the mod can connect to it for searching.

The relevant parts from my sphinx.conf:

You will need to change *AT LEAST* the database settings, i.e. username, password, database-name and maybe
the table name for the messages table used.
If you change the index names you will need to change them in the sphinx_search too.

ATTENTION:

The delta indexes require MySQL-4.1 or higher because of the use of subqueries and they need a separate table
for holding the last indexed ID, its structure is the following (install in phorum database). see sph_counter.sql file.

For Sample (working) sphinx config see sphinx.conf. Make sure to AT LEAST change the database settings.

KNOWN ISSUES:
It is not currently possible to search by author and by messages at the same
time. Search is performed either from phorum5_msg or phorum5_author sphinx
indexes.


Changelog
Changelog:
----------
2013-04-14 v1.1.2 by Elan Ruusamäe <glen@delfi.ee>
	- Restore original Phorum compatibility

2012-09-03 v1.1.1 by Elan Ruusamäe <glen@delfi.ee>
	- Allow users to search by author when they leave search keyword empty. PR#1.

2010-07-25 v1.1 by Elan Ruusamäe <glen@delfi.ee>
	- Update to work with recent phorum 5.2.14a and Sphinx 0.9.9
	- Added support for using sphinx pure PHP API and PHP PECL API
	- More complete crontab, sql and config files
	- Added info how to avoid filling original MySQL search database

2009-03-05 v1.0.0 by Martijn van Maasakkers (martijn@vanmaasakkers.net)
	- Mar 05, 2009
	- Update version for use with PHORUM 5.2
	- Added settings page where you can set the hostname and port of the sphinx daemon

2007-04-24 v0.9.2 by Thomas Seifert (thomas@phorum.org)
	- limited results to 1000
	- some security for the forum_id

2006-12-02 v0.9.1 by Thomas Seifert (thomas@phorum.org)
	- Updated to sphinx 0.9.7, added delta indexes and a host-ip

2006-10-13 v0.9 by Maurice Makaay (maurice@phorum.org)
	- Return search control to the Phorum search script in case an
	  unhandled type of search is encountered (e.g. the USER_ID search
	  type for Phorum 5.2).

2006-09-13 v0.8 by Maurice Makaay (maurice@phorum.org)
	- little hack in SetGroups of the sphinxapi.php script to not deny
	  group 0 (is_int(0) is false, but we can have 0 for announcements)
	- if no results are found, return empty results immediately (before,
	  no results would trigger undefined index warnings)
	- changed example config in README to use settings inheritance

2006-09-07 v0.7
	- received some more testing and hence ...
	- fix for date-order

2006-07-13 v0.5
	- first implementation, without admin interface


settings.php
<?php

if (!defined("PHORUM_ADMIN")) return;
if (!defined('PHORUM_INCLUDES_DIR')) {
	define('PHORUM_INCLUDES_DIR', './include'):
}

require_once 'defaults.php';

// save settings
if (count($_POST)) {
	$PHORUM['mod_sphinx_search'] = array(
		'hostname'  => $_POST['hostname'],
		'port'      => $_POST['port'],
	);

	if (!phorum_db_update_settings(array('mod_sphinx_search' => $PHORUM['mod_sphinx_search']))) {
		phorum_admin_error("Updating the settings in the database failed.");
	} else {
		phorum_admin_okmsg("Settings updated");
	}
}

?>
<div style="font-size: xx-large; font-weight: bold">Sphinx Search Module</div>
 This module uses the sphinx fulltext search engine to gather the results of the phorum-search.<br />
 On this page you can set the hostname and port of your sphinx search daemon.

<br style="clear:both" />
<?php

include_once PHORUM_INCLUDES_DIR.'/admin/PhorumInputForm.php';
$frm = new PhorumInputForm ("", "post", "Save");
$frm->hidden("module", "modsettings");
$frm->hidden("mod", "sphinx_search");

$frm->addbreak("Hostname and port");

$row = $frm->addrow("What is the hostname of the sphinx daemon? (e.g. 127.0.0.1)$warn", $frm->text_box("hostname", $PHORUM["mod_sphinx_search"]["hostname"], 30));
$row = $frm->addrow("What is the port of the sphinx daemon? (e.g. 9312)$warn", $frm->text_box("port", $PHORUM["mod_sphinx_search"]["port"], 30));
$frm->show();


sphinxclient.php
<?php

//
// $Id: sphinxapi.php,v 1.37 2006/11/26 12:24:21 shodan Exp $
//

//
// Copyright (c) 2001-2006, Andrew Aksyonoff. All rights reserved.
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License. You should have
// received a copy of the GPL license along with this program; if you
// did not, you can find it at http://www.gnu.org/
//

/////////////////////////////////////////////////////////////////////////////
// PHP version of Sphinx searchd client (PHP API)
/////////////////////////////////////////////////////////////////////////////

/// known searchd commands
define ( "SEARCHD_COMMAND_SEARCH",	0 );
define ( "SEARCHD_COMMAND_EXCERPT",	1 );

/// current client-side command implementation versions
define ( "VER_COMMAND_SEARCH",		0x104 );
define ( "VER_COMMAND_EXCERPT",		0x100 );

/// known searchd status codes
define ( "SEARCHD_OK",				0 );
define ( "SEARCHD_ERROR",			1 );
define ( "SEARCHD_RETRY",			2 );

/// known match modes
define ( "SPH_MATCH_ALL",			0 );
define ( "SPH_MATCH_ANY",			1 );
define ( "SPH_MATCH_PHRASE",		2 );
define ( "SPH_MATCH_BOOLEAN",		3 );

/// known sort modes
define ( "SPH_SORT_RELEVANCE",		0 );
define ( "SPH_SORT_ATTR_DESC",		1 );
define ( "SPH_SORT_ATTR_ASC",		2 );
define ( "SPH_SORT_TIME_SEGMENTS", 	3 );
define ( "SPH_SORT_EXTENDED", 		4 );

/// known attribute types
define ( "SPH_ATTR_INTEGER",		1 );
define ( "SPH_ATTR_TIMESTAMP",		2 );

/// known grouping functions
define ( "SPH_GROUPBY_DAY",			0 );
define ( "SPH_GROUPBY_WEEK",		1 );
define ( "SPH_GROUPBY_MONTH",		2 );
define ( "SPH_GROUPBY_YEAR",		3 );
define ( "SPH_GROUPBY_ATTR",		4 );

/// sphinx searchd client class
class SphinxClient
{
	var $_host;		///< searchd host (default is "localhost")
	var $_port;		///< searchd port (default is 3312)
	var $_offset;	///< how many records to seek from result-set start (default is 0)
	var $_limit;	///< how many records to return from result-set starting at offset (default is 20)
	var $_mode;		///< query matching mode (default is SPH_MATCH_ALL)
	var $_weights;	///< per-field weights (default is 1 for all fields)
	var $_sort;		///< match sorting mode (default is SPH_SORT_RELEVANCE)
	var $_sortby;	///< attribute to sort by (defualt is "")
	var $_min_id;	///< min ID to match (default is 0)
	var $_max_id;	///< max ID to match (default is UINT_MAX)
	var $_min;		///< attribute name to min-value hash (for range filters)
	var $_max;		///< attribute name to max-value hash (for range filters)
	var $_filter;	///< attribute name to values set hash (for values-set filters)
	var $_groupby;	///< group-by attribute name
	var $_groupfunc;///< function to pre-process group-by attribute value with
	var $_maxmatches;///< max matches to retrieve

	var $_error;	///< last error message
	var $_warning;	///< last warning message

	/////////////////////////////////////////////////////////////////////////////
	// common stuff
	/////////////////////////////////////////////////////////////////////////////

	/// create a new client object and fill defaults
	function SphinxClient ()
	{
		$this->_host	= "localhost";
		$this->_port	= 3312;
		$this->_offset	= 0;
		$this->_limit	= 20;
		$this->_mode	= SPH_MATCH_ALL;
		$this->_weights	= array ();
		$this->_sort	= SPH_SORT_RELEVANCE;
		$this->_sortby	= "";
		$this->_min_id	= 0;
		$this->_max_id	= 0xFFFFFFFF;
		$this->_min		= array ();
		$this->_max		= array ();
		$this->_filter	= array ();
		$this->_groupby		= "";
		$this->_groupfunc	= SPH_GROUPBY_DAY;
		$this->_maxmatches	= 1000;

		$this->_error	= "";
		$this->_warning	= "";
	}

	/// get last error message (string)
	function GetLastError ()
	{
		return $this->_error;
	}

	/// get last warning message (string)
	function GetLastWarning ()
	{
		return $this->_warning;
	}

	/// set searchd server
	function SetServer ( $host, $port )
	{
		assert ( is_string($host) );
		assert ( is_int($port) );
		$this->_host = $host;
		$this->_port = $port;
	}

	/////////////////////////////////////////////////////////////////////////////

	/// connect to searchd server
	function _Connect ()
	{
		if (!( $fp = @fsockopen ( $this->_host, $this->_port ) ) )
		{
			$this->_error = "connection to {$this->_host}:{$this->_port} failed";
			return false;
		}

		// check version
		list(,$v) = unpack ( "N*", fread ( $fp, 4 ) );
		$v = (int)$v;
		if ( $v<1 )
		{
			fclose ( $fp );
			$this->_error = "expected searchd protocol version 1+, got version '$v'";
			return false;
		}

		// all ok, send my version
		fwrite ( $fp, pack ( "N", 1 ) );
		return $fp;
	}

	/// get and check response packet from searchd server
	function _GetResponse ( $fp, $client_ver )
	{
		$header = fread ( $fp, 8 );
		list ( $status, $ver, $len ) = array_values ( unpack ( "n2a/Nb", $header ) );
		$response = "";
		$left = $len;
		while ( $left>0 && !feof($fp) )
		{
			$chunk = fread ( $fp, $left );
			if ( $chunk )
			{
				$response .= $chunk;
				$left -= strlen($chunk);
			}
		}
		fclose ( $fp );

		// check response
		$read = strlen ( $response );
		if ( !$response || $read!=$len )
		{
			$this->_error = $len
				? "failed to read searchd response (status=$status, ver=$ver, len=$len, read=$read)"
				: "received zero-sized searchd response";
			return false;
		}

		// check status
		if ( $status==SEARCHD_ERROR )
		{
			$this->_error = "searchd error: " . substr ( $response, 4 );
			return false;
		}
		if ( $status==SEARCHD_RETRY )
		{
			$this->_error = "temporary searchd error: " . substr ( $response, 4 );
			return false;
		}
		if ( $status!=SEARCHD_OK )
		{
			$this->_error = "unknown status code '$status'";
			return false;
		}

		// check version
		if ( $ver<$client_ver )
		{
			$this->_warning = sprintf ( "searchd command v.%d.%d older than client's v.%d.%d, some options might not work",
				$ver>>8, $ver&0xff, $client_ver>>8, $client_ver&0xff );
		}

		return $response;
	}

	/////////////////////////////////////////////////////////////////////////////
	// searching
	/////////////////////////////////////////////////////////////////////////////

	/// set match offset, count, and max number to retrieve
	function SetLimits ( $offset, $limit, $max=0 )
	{
		assert ( is_int($offset) );
		assert ( is_int($limit) );
		assert ( $offset>=0 );
		assert ( $limit>0 );
		assert ( $max>=0 );
		$this->_offset = $offset;
		$this->_limit = $limit;
		if ( $max>0 )
			$this->_maxmatches = $max;
	}

	/// set match mode
	function SetMatchMode ( $mode )
	{
		assert ( $mode==SPH_MATCH_ALL
			|| $mode==SPH_MATCH_ANY
			|| $mode==SPH_MATCH_PHRASE
			|| $mode==SPH_MATCH_BOOLEAN );
		$this->_mode = $mode;
	}

	/// set sort mode
	function SetSortMode ( $mode, $sortby="" )
	{
		assert (
			$mode==SPH_SORT_RELEVANCE ||
			$mode==SPH_SORT_ATTR_DESC ||
			$mode==SPH_SORT_ATTR_ASC ||
			$mode==SPH_SORT_TIME_SEGMENTS ||
			$mode==SPH_SORT_EXTENDED );
		assert ( is_string($sortby) );
		assert ( $mode==SPH_SORT_RELEVANCE || strlen($sortby)>0 );

		$this->_sort = $mode;
		$this->_sortby = $sortby;
	}

	/// set per-field weights
	function SetWeights ( $weights )
	{
		assert ( is_array($weights) );
		foreach ( $weights as $weight )
			assert ( is_int($weight) );

		$this->_weights = $weights;
	}

	/// set IDs range to match
	/// only match those records where document ID
	/// is beetwen $min and $max (including $min and $max)
	function SetIDRange ( $min, $max )
	{
		assert ( is_int($min) );
		assert ( is_int($max) );
		assert ( $min<=$max );
		$this->_min_id = $min;
		$this->_max_id = $max;
	}

	/// set values filter
	/// only match those records where $attribute column values
	/// are in specified set
	function SetFilter ( $attribute, $values )
	{
		assert ( is_string($attribute) );
		assert ( is_array($values) );
		assert ( count($values) );

		if ( is_array($values) && count($values) )
		{
			foreach ( $values as $value )
				assert ( is_int($value) );

			$this->_filter[$attribute] = $values;
		}
	}

	/// set range filter
	/// only match those records where $attribute column value
	/// is beetwen $min and $max (including $min and $max)
	function SetFilterRange ( $attribute, $min, $max )
	{
		assert ( is_string($attribute) );
		assert ( is_int($min) );
		assert ( is_int($max) );
		assert ( $min<=$max );

		$this->_min[$attribute] = $min;
		$this->_max[$attribute] = $max;
	}

	/// set grouping attribute and function
	///
	/// in grouping mode, all matches are assigned to different groups
	/// based on grouping function value.
	///
	/// each group keeps track of the total match count, and the best match
	/// (in this group) according to current sorting function.
	///
	/// the final result set contains one best match per group, with
	/// grouping function value and matches count attached. result set
	/// is sorted by grouping function value, in descending order.
	///
	/// for example, if sorting by relevance and grouping by "published"
	/// attribute with SPH_GROUPBY_DAY function, then the result set will
	/// contain one most relevant match per each day when there were any
	/// matches published, with day number and per-day match count attached,
	/// and sorted by day number in descending order (ie. recent days first).
	function SetGroupBy ( $attribute, $func )
	{
		assert ( is_string($attribute) );
		assert ( $func==SPH_GROUPBY_DAY
			|| $func==SPH_GROUPBY_WEEK
			|| $func==SPH_GROUPBY_MONTH
			|| $func==SPH_GROUPBY_YEAR
			|| $func==SPH_GROUPBY_ATTR );

		$this->_groupby = $attribute;
		$this->_groupfunc = $func;
	}

	/// connect to searchd server and run given search query
	///
	/// $query is query string
	/// $index is index name to query, default is "*" which means to query all indexes
	///
	/// returns false on failure
	/// returns hash which has the following keys on success:
	///		"matches"
	///			hash which maps found document_id to ( "weight", "group" ) hash
	///		"total"
	///			total amount of matches retrieved (upto SPH_MAX_MATCHES, see sphinx.h)
	///		"total_found"
	///			total amount of matching documents in index
	///		"time"
	///			search time
	///		"words"
	///			hash which maps query terms (stemmed!) to ( "docs", "hits" ) hash
	function Query ( $query, $index="*" )
	{
		if (!( $fp = $this->_Connect() ))
			return false;

		/////////////////
		// build request
		/////////////////

		$req = pack ( "NNNN", $this->_offset, $this->_limit, $this->_mode, $this->_sort ); // mode and limits
		$req .= pack ( "N", strlen($this->_sortby) ) . $this->_sortby;
		$req .= pack ( "N", strlen($query) ) . $query; // query itself
		$req .= pack ( "N", count($this->_weights) ); // weights
		foreach ( $this->_weights as $weight )
			$req .= pack ( "N", (int)$weight );
		$req .= pack ( "N", strlen($index) ) . $index; // indexes
		$req .= // id range
			pack ( "N", (int)$this->_min_id ) .
			pack ( "N", (int)$this->_max_id );

		// filters
		$req .= pack ( "N", count($this->_min) + count($this->_filter) );

		foreach ( $this->_min as $attr => $min )
			$req .=
				pack ( "N", strlen($attr) ) . $attr .
				pack ( "NNN", 0, $min, $this->_max[$attr] );

		foreach ( $this->_filter as $attr => $values )
		{
			$req .= 
				pack ( "N", strlen($attr) ) . $attr .
				pack ( "N", count($values) );

			foreach ( $values as $value )
				$req .= pack ( "N", $value );
		}

		// group-by
		$req .= pack ( "NN", $this->_groupfunc, strlen($this->_groupby) ) . $this->_groupby;

		// max matches to retrieve
		$req .= pack ( "N", $this->_maxmatches );

		////////////////////////////
		// send query, get response
		////////////////////////////

		$len = strlen($req);
		$req = pack ( "nnN", SEARCHD_COMMAND_SEARCH, VER_COMMAND_SEARCH, $len ) . $req; // add header
		fwrite ( $fp, $req, $len+8 );
		if (!( $response = $this->_GetResponse ( $fp, VER_COMMAND_SEARCH ) ))
			return false;

		//////////////////
		// parse response
		//////////////////

		$result = array();
		$max = strlen($response); // protection from broken response

		// read schema
		$p = 0;
		$fields = array ();
		$attrs = array ();

		list(,$nfields) = unpack ( "N*", substr ( $response, $p, 4 ) ); $p += 4;
		while ( $nfields-->0 && $p<$max )
		{
			list(,$len) = unpack ( "N*", substr ( $response, $p, 4 ) ); $p += 4;
			$fields[] = substr ( $response, $p, $len ); $p += $len;
		}
		$result["fields"] = $fields;

		list(,$nattrs) = unpack ( "N*", substr ( $response, $p, 4 ) ); $p += 4;
		while ( $nattrs-->0 && $p<$max  )
		{
			list(,$len) = unpack ( "N*", substr ( $response, $p, 4 ) ); $p += 4;
			$attr = substr ( $response, $p, $len ); $p += $len;
			list(,$type) = unpack ( "N*", substr ( $response, $p, 4 ) ); $p += 4;
			$attrs[$attr] = $type;
		}
		$result["attrs"] = $attrs;

		// read match count
		list(,$count) = unpack ( "N*", substr ( $response, $p, 4 ) ); $p += 4;

		// read matches
		while ( $count-->0 && $p<$max )
		{
			list ( $doc, $weight ) = array_values ( unpack ( "N*N*",
				substr ( $response, $p, 8 ) ) );
			$p += 8;

			$result["matches"][$doc]["weight"] = $weight;
			foreach ( $attrs as $attr=>$type )
			{
				list(,$val) = unpack ( "N*", substr ( $response, $p, 4 ) ); $p += 4;
				$result["matches"][$doc]["attrs"][$attr] = $val;
			}
		}
		list ( $result["total"], $result["total_found"], $result["time"], $words ) =
			array_values ( unpack ( "N*N*N*N*", substr ( $response, $p, 16 ) ) );
		$result["time"] = sprintf ( "%.3f", $result["time"]/1000 );
		$p += 16;

		while ( $words-->0 )
		{
			list(,$len) = unpack ( "N*", substr ( $response, $p, 4 ) ); $p += 4;
			$word = substr ( $response, $p, $len ); $p += $len;
			list ( $docs, $hits ) = array_values ( unpack ( "N*N*", substr ( $response, $p, 8 ) ) ); $p += 8;

			$result["words"][$word] = array ( "docs"=>$docs, "hits"=>$hits );
		}

		return $result;
	}

	/////////////////////////////////////////////////////////////////////////////
	// excerpts generation
	/////////////////////////////////////////////////////////////////////////////

	/// connect to searchd server and generate exceprts from given documents
	///
	/// $docs is an array of strings which represent the documents' contents
	/// $index is a string specifiying the index which settings will be used
	///		for stemming, lexing and case folding
	/// $words is a string which contains the words to highlight
	/// $opts is a hash which contains additional optional highlighting parameters:
	///		"before_match"
	///			a string to insert before a set of matching words, default is "<b>"
	///		"after_match"
	///			a string to insert after a set of matching words, default is "<b>"
	///		"chunk_separator"
	///			a string to insert between excerpts chunks, default is " ... "
	///		"limit"
	///			max excerpt size in symbols (codepoints), default is 256
	///		"around"
	///			how much words to highlight around each match, default is 5
	///
	/// returns false on failure
	/// returns an array of string excerpts on success
	function BuildExcerpts ( $docs, $index, $words, $opts=array() )
	{
		assert ( is_array($docs) );
		assert ( is_string($index) );
		assert ( is_string($words) );
		assert ( is_array($opts) );

		if (!( $fp = $this->_Connect() ))
			return false;

		/////////////////
		// fixup options
		/////////////////

		if ( !isset($opts["before_match"]) )		$opts["before_match"] = "<b>";
		if ( !isset($opts["after_match"]) )			$opts["after_match"] = "</b>";
		if ( !isset($opts["chunk_separator"]) )		$opts["chunk_separator"] = " ... ";
		if ( !isset($opts["limit"]) )				$opts["limit"] = 256;
		if ( !isset($opts["around"]) )				$opts["around"] = 5;

		/////////////////
		// build request
		/////////////////

		// v.1.0 req
		$req = pack ( "NN", 0, 1 ); // mode=0, flags=1 (remove spaces)
		$req .= pack ( "N", strlen($index) ) . $index; // req index
		$req .= pack ( "N", strlen($words) ) . $words; // req words

		// options
		$req .= pack ( "N", strlen($opts["before_match"]) ) . $opts["before_match"];
		$req .= pack ( "N", strlen($opts["after_match"]) ) . $opts["after_match"];
		$req .= pack ( "N", strlen($opts["chunk_separator"]) ) . $opts["chunk_separator"];
		$req .= pack ( "N", (int)$opts["limit"] );
		$req .= pack ( "N", (int)$opts["around"] );

		// documents
		$req .= pack ( "N", count($docs) );
		foreach ( $docs as $doc )
		{
			assert ( is_string($doc) );
			$req .= pack ( "N", strlen($doc) ) . $doc;
		}

		////////////////////////////
		// send query, get response
		////////////////////////////

		$len = strlen($req);
		$req = pack ( "nnN", SEARCHD_COMMAND_EXCERPT, VER_COMMAND_EXCERPT, $len ) . $req; // add header
		$wrote = fwrite ( $fp, $req, $len+8 );
		if (!( $response = $this->_GetResponse ( $fp, VER_COMMAND_EXCERPT ) ))
			return false;

		//////////////////
		// parse response
		//////////////////

		$pos = 0;
		$res = array ();
		$rlen = strlen($response);
		for ( $i=0; $i<count($docs); $i++ )
		{
			list(,$len) = unpack ( "N*", substr ( $response, $pos, 4 ) );
			$pos += 4;

			if ( $pos+$len > $rlen )
			{
				$this->_error = "incomplete reply";
				return false;
			}
			$res[] = substr ( $response, $pos, $len );
			$pos += $len;
		}

		return $res;
	}
}

//
// $Id: sphinxapi.php,v 1.37 2006/11/26 12:24:21 shodan Exp $
//

?>

sphinx_search.php
<?php

/*
 * Module for using sphinx as fulltext search-engine
 * sphinx can be downloaded and installed from
 * http://sphinxsearch.com/
 *
 * mod written by Thomas Seifert (thomas@phorum.org)
 * fixes/enhancements by Maurice Makaay (maurice@phorum.org)
 * update for Phorum 5.2 by Martijn van Maasakkers (martijn@vanmaasakkers.net)
 * more fixes/improvements by Elan Ruusamäe <glen@delfi.ee>
 *
 * See Changelog for version history
 *
 */

if (!defined('PHORUM')) return;

require_once 'defaults.php';

function sphinx_search_action($arrSearch)
{
	global $PHORUM;

	// No pecl class, try php version
	if (!class_exists('SphinxClient')) {
		// loads from php include_path
		require_once 'sphinxapi.php';
	}

	// these are the index-names set in sphinx.conf - one for searching messages, the other for searching by authors only
	// both contain an additional index for the deltas - changes done after the last full reindex
	$index_name_msg = 'phorum5_msg_d phorum5_msg';
	$index_name_author = 'phorum5_author phorum5_author_d';

	// excerpts_index is just one index as that function only accepts one, it used for determining charsets / mapping tables, nothing more
	$excerpts_index = 'phorum5_msg';

	$index = $index_name_msg;

	if ($arrSearch['match_type'] == 'ALL') {
			$match_mode = SPH_MATCH_ALL;
	} elseif($arrSearch['match_type'] == 'ANY') {
			$match_mode = SPH_MATCH_ANY;
	} elseif($arrSearch['match_type'] == 'PHRASE') {
			$match_mode = SPH_MATCH_PHRASE;
	} elseif($arrSearch['match_type'] == 'AUTHOR') {
			$match_mode = SPH_MATCH_PHRASE;
			$index = $index_name_author;
	} else {
			// Return search control to Phorum in case the search type isn't handled by the module.
			return $arrSearch;
	}

	if (empty($arrSearch['search']) && !empty($arrSearch['author'])) {
		$arrSearch['search'] = $arrSearch['author'];
		$index = $index_name_author;
	}

	$sphinx = new SphinxClient();
	$sphinx->SetServer($PHORUM['mod_sphinx_search']['hostname'], $PHORUM['mod_sphinx_search']['port']);
	$sphinx->SetMatchMode($match_mode);

	// set the limits for paging
	$sphinx->SetLimits($arrSearch['offset'],$arrSearch['length']);

	// set the timeframe to search
	if ($arrSearch['match_dates'] > 0) {
		$min_ts = time() - 86400 * $arrSearch['match_dates'];
		$max_ts = time();
		$sphinx->SetFilterRange('datestamp', $min_ts, $max_ts);

	}

	// Check what forums the active Phorum user can read.
	$allowed_forums = phorum_api_user_check_access(
		PHORUM_USER_ALLOW_READ, PHORUM_ACCESS_LIST
	);

	// If the user is not allowed to search any forum or the current
	// active forum, then return the emtpy search results array.
	if (empty($allowed_forums) || ($PHORUM['forum_id'] > 0 && !in_array($PHORUM['forum_id'], $allowed_forums))) {
		$arrSearch['results'] = array();
		$arrSearch['totals'] = 0;
		$arrSearch['continue'] = 0;
		$arrSearch['raw_body'] = 1;
		return $arrSearch;
	}

	// Prepare forum_id restriction.
	$search_forums = array();
	foreach (explode(',', $arrSearch['match_forum']) as $forum_id) {
		if ($forum_id == 'ALL') {
			$search_forums = $allowed_forums;
			break;
		}
		if (isset($allowed_forums[$forum_id])) {
			$search_forums[] = $forum_id;
		}
	}
	$sphinx->SetFilter('forum_id', $search_forums);

	// set the sort-mode
	$sphinx->SetSortMode(SPH_SORT_ATTR_DESC, 'datestamp');

	// do the actual query
	$results = $sphinx->Query($arrSearch['search'], $index);

	$res = $sphinx->GetLastWarning();
	if ($res) {
		error_log("sphinx_search.php: WARNING: $res");
	}
	$res = $sphinx->GetLastError();
	if ($res) {
		error_log("sphinx_search.php: ERROR: $res");
	}

	// if no messages were found, then return empty handed.
	if (!isset($results['matches'])) {
		$arrSearch['results']=array();
		$arrSearch['totals']=0;
		$arrSearch['continue']=0;
		$arrSearch['raw_body']=1;
		return $arrSearch;
	}

	$search_msg_ids = $results['matches'];

	// get the messages we found
	$found_messages = phorum_db_get_message(array_keys($search_msg_ids), 'message_id', true);

	// sort them in reverse order of the message_id to automagically sort them by date desc this way
	krsort($found_messages);
	reset($found_messages);

	// prepare the array for building highlighted excerpts
	$docs = array();
	foreach($found_messages as $id => $data) {
		// remove hidden text in the output - only added by the hidden_msg module
		$data['body']=preg_replace("/(\[hide=([\#a-z0-9]+?)\](.+?)\[\/hide\])/is", '', $data['body']);

		$docs[] = htmlspecialchars(phorum_strip_body($data['body']));
	}

	$words = '';
	if (!empty($results['words'])) {
		$words = implode(' ', array_keys($results['words']));
	}

	$opts = array('chunk_separator'=>' [...] ');

	// build highlighted excerpts
	$highlighted = $sphinx->BuildExcerpts($docs,$excerpts_index,$words,$opts);

	$res = $sphinx->GetLastWarning();
	if ($res) {
		error_log("sphinx_search.php: WARNING: $res");
	}
	$res = $sphinx->GetLastError();
	if ($res) {
		error_log("sphinx_search.php: ERROR: $res");
	}

	$cnt=0;
	foreach ($found_messages as $id => $content) {
		$found_messages[$id]['short_body'] = $highlighted[$cnt];
		$cnt++;
	}

	$arrSearch['results'] = $found_messages;
	// we need the total results
	$arrSearch['totals'] = $results['total_found'];

	if ($arrSearch['totals'] > 1000) {
		$arrSearch['totals'] = 1000;
	}

	// don't run the default search
	$arrSearch['continue'] = 0;
	// tell it to leave the body alone
	$arrSearch['raw_body'] = 1;

	return $arrSearch;
}
